homeassistant/components/legrand_rflc/config_flow.py

```diff
@@ -0,0 +1,171 @@
+            task = self.hass.async_create_task(
+                lc7001.aio.Connector(host, **kwargs).loop()
+            )
+            try:
+                mac = await task
+            except OSError:
+                errors[CONF_HOST] = self.ERROR_INVALID_HOST
+            except lc7001.aio.Authenticator.Error:
+                errors[CONF_PASSWORD] = self.ERROR_INVALID_AUTH
+            else:
+                data = {CONF_HOST: host, CONF_MAC: mac}
+                if key is not None:
+                    data[CONF_AUTHENTICATION] = key.hex()
+                return self.async_create_entry(title=host, data=data)
+
+        # get user_input
+        return self.async_show_form(
+            step_id="user",
+            data_schema=voluptuous.Schema(
+                {
+                    voluptuous.Required(CONF_HOST, default=host): str,
+                    voluptuous.Optional(CONF_PASSWORD): str,
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_reauth(
+        self, user_input: dict[str, Any] | None = None
+    ) -> data_entry_flow.FlowResult:
+        """Handle configuration by reauth."""
+        host = self.context["unique_id"]
+        errors = {CONF_PASSWORD: self.ERROR_INVALID_AUTH}
+        if user_input is not None:
+            key: bytes | None = None
+            kwargs = {"key": key, "loop_timeout": -1}
+            if CONF_AUTHENTICATION in user_input:
+                key = kwargs["key"] = bytes.fromhex(user_input[CONF_AUTHENTICATION])
+            if CONF_PORT in user_input:  # for testing server emulation on localhost
+                kwargs["port"] = user_input[CONF_PORT]
+            if CONF_PASSWORD in user_input:
+                key = kwargs["key"] = lc7001.aio.hash_password(
+                    user_input[CONF_PASSWORD].encode()
+                )
+            task = self.hass.async_create_task(
+                lc7001.aio.Connector(host, **kwargs).loop()
+            )
+            try:
+                mac = await task
+            except OSError:
+                errors[CONF_HOST] = self.ERROR_INVALID_HOST
+            except lc7001.aio.Authenticator.Error:
+                pass
+            else:
+                data = {CONF_HOST: host, CONF_MAC: mac}
+                if CONF_PORT in user_input:  # for testing server emulation on localhost
+                    data[CONF_PORT] = user_input[CONF_PORT]
+                if key is not None:
+                    data[CONF_AUTHENTICATION] = key.hex()
+                entry = self.context["entry"]
+                self.hass.config_entries.async_update_entry(entry, data=data)
+                self.hass.async_create_task(
+                    self.hass.config_entries.async_setup(entry.entry_id)
+                )
+                return self.async_abort(reason=self.ABORT_REAUTH_SUCCESSFUL)
+
+        # get user_input
+        return self.async_show_form(
+            step_id="reauth",
```

Comment:
async_step_reauth is invoked whenever interaction with a configured lc7001 hub fails to authenticate. This can happen any time a (re)connection is made to the hub. Before it is called, the ConfigEntry for it is removed (unloaded) for the hub.

Initially, the old authentication data (that worked before) is passed in and sometimes this works (no user interaction required). Only if authentication fails is a form shown to the user. This form allows the user to change the host and the password. We go round and round with this until a host/password combination works. When it works a new config entry is setup.

I don't see how/where a confirmation reauth step fits into this flow.

Reply:
A reauth must always have at least a confirmation from the user before finishing.

The old config entry should not be removed, just unloaded. The reauth flow should update the old config entry with the new data and abort with success message and reload the config entry.

When using our helper tools to start a reauth flow, which we should do, the existing entry data is passed as input to the reauth step. Add a confirmation step that shows the form to the user.

Reply:
We should only start a reauth flow if the old credentials fail. We can try again in the integration with the existing credentials before starting a reauth flow. If they fail, then we start a reauth flow and ask the user to enter new credentials.

Reply:
Did you see my comment on the related issue?

More than once, I have seen my hub fail to authenticate even when I have given it the proper credentials. So, in case this has happened, I (just) unload (not remove) the config_entry and start the reauth flow. As it is now, this flow may succeed using these old credentials (I have seen it happen on my hub). In this case, there is no need to interact with the user - the system automatically heals itself. When this does not work, the user is asked to confirm using the same async_step_reauth method. I don't see how I can make this better.

> We should only start a reauth flow if the old credentials fail. We can try again in the integration with the existing credentials before starting a reauth flow. If they fail, then we start a reauth flow and ask the user to enter new credentials.

I do **all** of this in one place (DNRY): async_step_reauth. I do not see how doing a re-auth before async_step_reauth is better than doing the reauth only in async_step_reauth and doing the confirmation (if it fails) there as well.

Reply:
If there's no need to interact with the user, we don't need to start a reauth flow. Just reload the integration or whatever is needed to authenticate again automatically.

Reply:
> If there's no need to interact with the user, we don't need to start a reauth flow

What is the harm? The code is much simpler this way. It is all in one place. No need for another config flow step.

> Just reload the integration or whatever is needed to authenticate again automatically.

Eventually, I will have to handle the failure. Changing it from the way it is now will introduce new/more code that needs to be tested and maintained with no real benefit.

Reply:
I suggest reloading the integration if encountering an auth failure after config entry is setup. This will try to authenticate again during the normal config entry setup, and if we then fail we start the reauth flow. This will give you one attempt to login again with existing credentials, and virtually no extra code. It's just one line to reload the config entry.

Reply:
> I suggest reloading the integration if encountering an auth failure after config entry is setup.

Since this failure occurs asynchronously, this will require the introduction of state information. When the failure occurs, the context in which it occurs will need to be evaluated to determine whether to reload or unload-and-reauth. There is no need to do any of this. The code is much simpler and performs the exact same function as it is now.

Reply:
https://github.com/home-assistant/core/pull/53430/commits/b2afb8da7bb8f4ce0b2c0c113ab7624f36df5693